File: src/ros_template_node/src/structured_logger.cpp
#include "ros_template_node/structured_logger.hpp"
#include <iomanip>
#include <sstream>
#include <random>

namespace ros_template_node {

StructuredLogger::StructuredLogger(const std::string& node_name) 
    : node_name_(node_name), correlation_id_(generate_correlation_id()) {
    
    // Create console logger with JSON format
    logger_ = spdlog::stdout_color_mt(node_name + "_structured");
    logger_->set_pattern("%v");  // Only print the message (our JSON)
    logger_->set_level(spdlog::level::debug);
}

void StructuredLogger::log(Level level, Component component, EventType event_type,
                          const std::string& message, const nlohmann::json& context) {
    nlohmann::json log_entry;
    
    log_entry["timestamp"] = get_iso_timestamp();
    log_entry["level"] = level_to_string(level);
    log_entry["node_name"] = node_name_;
    log_entry["component"] = component_to_string(component);
    log_entry["event_type"] = event_type_to_string(event_type);
    log_entry["message"] = message;
    log_entry["context"] = context;
    log_entry["correlation_id"] = correlation_id_;

    // Convert to string and log
    std::string json_str = log_entry.dump();
    
    switch (level) {
        case Level::DEBUG:
            logger_->debug(json_str);
            break;
        case Level::INFO:
            logger_->info(json_str);
            break;
        case Level::WARN:
            logger_->warn(json_str);
            break;
        case Level::ERROR:
            logger_->error(json_str);
            break;
        case Level::FATAL:
            logger_->critical(json_str);
            break;
    }
}

void StructuredLogger::debug(Component component, EventType event_type,
                            const std::string& message, const nlohmann::json& context) {
    log(Level::DEBUG, component, event_type, message, context);
}

void StructuredLogger::info(Component component, EventType event_type,
                           const std::string& message, const nlohmann::json& context) {
    log(Level::INFO, component, event_type, message, context);
}

void StructuredLogger::warn(Component component, EventType event_type,
                           const std::string& message, const nlohmann::json& context) {
    log(Level::WARN, component, event_type, message, context);
}

void StructuredLogger::error(Component component, EventType event_type,
                            const std::string& message, const nlohmann::json& context) {
    log(Level::ERROR, component, event_type, message, context);
}

std::string StructuredLogger::level_to_string(Level level) {
    switch (level) {
        case Level::DEBUG: return "DEBUG";
        case Level::INFO: return "INFO";
        case Level::WARN: return "WARN";
        case Level::ERROR: return "ERROR";
        case Level::FATAL: return "FATAL";
        default: return "UNKNOWN";
    }
}

std::string StructuredLogger::component_to_string(Component component) {
    switch (component) {
        case Component::STARTUP: return "startup";
        case Component::TIMER: return "timer";
        case Component::SUBSCRIBER: return "subscriber";
        case Component::HEALTH: return "health";
        case Component::SHUTDOWN: return "shutdown";
        default: return "unknown";
    }
}

std::string StructuredLogger::event_type_to_string(EventType event_type) {
    switch (event_type) {
        case EventType::PUBLISH: return "publish";
        case EventType::RECEIVE: return "receive";
        case EventType::HEALTH_CHECK: return "health_check";
        case EventType::INITIALIZATION: return "initialization";
        case EventType::PARAMETER_UPDATE: return "parameter_update";
        case EventType::ERROR_OCCURRED: return "error";
        default: return "unknown";
    }
}

std::string StructuredLogger::generate_correlation_id() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<> dis(0, 15);
    
    std::stringstream ss;
    for (int i = 0; i < 8; ++i) {
        ss << std::hex << dis(gen);
    }
    return ss.str();
}

std::string StructuredLogger::get_iso_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto time_t = std::chrono::system_clock::to_time_t(now);
    auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()) % 1000;
    
    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time_t), "%Y-%m-%dT%H:%M:%S");
    ss << '.' << std::setfill('0') << std::setw(3) << ms.count() << 'Z';
    return ss.str();
}

} // namespace ros_template_node

File: src/py_template_node/py_template_node/structured_logger.py
#!/usr/bin/env python3
"""Structured JSON logging for ROS 2 Python nodes."""

import json
import logging
import sys
import time
import uuid
from datetime import datetime, timezone
from enum import Enum
from typing import Any, Dict, Optional

import structlog


class LogLevel(Enum):
    """Log level enumeration."""

    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"
    FATAL = "FATAL"


class Component(Enum):
    """Component enumeration for structured logging."""

    STARTUP = "startup"
    TIMER = "timer"
    SUBSCRIBER = "subscriber"
    HEALTH = "health"
    SHUTDOWN = "shutdown"


class EventType(Enum):
    """Event type enumeration for structured logging."""

    PUBLISH = "publish"
    RECEIVE = "receive"
    HEALTH_CHECK = "health_check"
    INITIALIZATION = "initialization"
    PARAMETER_UPDATE = "parameter_update"
    ERROR_OCCURRED = "error"


class StructuredLogger:
    """Structured JSON logger for ROS 2 nodes."""

    def __init__(self, node_name: str):
        """Initialize the structured logger.

        Args:
            node_name: Name of the ROS 2 node
        """
        self.node_name = node_name
        self.correlation_id = str(uuid.uuid4())[:8]

        # Configure structlog
        structlog.configure(
            processors=[
                structlog.stdlib.filter_by_level,
                structlog.stdlib.add_logger_name,
                structlog.stdlib.add_log_level,
                structlog.stdlib.PositionalArgumentsFormatter(),
                structlog.processors.TimeStamper(fmt="iso"),
                structlog.processors.StackInfoRenderer(),
                structlog.processors.format_exc_info,
                structlog.processors.UnicodeDecoder(),
                structlog.processors.JSONRenderer(),
            ],
            context_class=dict,
            logger_factory=structlog.stdlib.LoggerFactory(),
            wrapper_class=structlog.stdlib.BoundLogger,
            cache_logger_on_first_use=True,
        )

        # Setup standard logging
        logging.basicConfig(
            format="%(message)s",
            stream=sys.stdout,
            level=logging.DEBUG,
        )

        self.logger = structlog.get_logger(node_name)

    def log(
        self,
        level: LogLevel,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log a structured message.

        Args:
            level: Log level
            component: Component generating the log
            event_type: Type of event being logged
            message: Human-readable message
            context: Additional context data
        """
        if context is None:
            context = {}

        log_data = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "level": level.value,
            "node_name": self.node_name,
            "component": component.value,
            "event_type": event_type.value,
            "message": message,
            "context": context,
            "correlation_id": self.correlation_id,
        }

        # Log at appropriate level
        if level == LogLevel.DEBUG:
            self.logger.debug("", **log_data)
        elif level == LogLevel.INFO:
            self.logger.info("", **log_data)
        elif level == LogLevel.WARN:
            self.logger.warning("", **log_data)
        elif level == LogLevel.ERROR:
            self.logger.error("", **log_data)
        elif level == LogLevel.FATAL:
            self.logger.critical("", **log_data)

    def debug(
        self,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log a debug message."""
        self.log(LogLevel.DEBUG, component, event_type, message, context)

    def info(
        self,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log an info message."""
        self.log(LogLevel.INFO, component, event_type, message, context)

    def warn(
        self,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log a warning message."""
        self.log(LogLevel.WARN, component, event_type, message, context)

    def error(
        self,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log an error message."""
        self.log(LogLevel.ERROR, component, event_type, message, context)

    def fatal(
        self,
        component: Component,
        event_type: EventType,
        message: str,
        context: Optional[Dict[str, Any]] = None,
    ) -> None:
        """Log a fatal message."""
        self.log(LogLevel.FATAL, component, event_type, message, context)
